tools: use local or specified $NODE for test-npm

Fishrock123 opened the pull request:
See: https://github.com/nodejs/io.js/pull/1955#discussion_r32291327

cc @nodejs/build

Fishrock123 commented on Makefile:
```diff
@@ -136,7 +136,7 @@ test-debugger: all
 	$(PYTHON) tools/test.py debugger
 
 test-npm: $(NODE_EXE)
```
These are still correct, right? I'm not too sure how makefile uses this part.

jbergstroem replied:
Yeah, NODE_EXE is mostly used as build target.

jbergstroem commented on tools/test-npm.sh:
```diff
@@ -29,8 +29,8 @@ export npm_config_prefix="npm-prefix"
-../$NODE_EXE cli.js install --ignore-scripts
-../$NODE_EXE cli.js run-script test-all
+../$NODE cli.js install --ignore-scripts
+../$NODE cli.js run-script test-all
```
Aren't we changing directory a few lines up? Thinking this should just be `$NODE`. Guess it depends on how you pass the binary path.

Fishrock123 replied:
It's not `./node` because at this point we are inside the `test-npm` directory. :)

jbergstroem replied:
Yeah, you're right

Fishrock123 commented:
@jbergstroem is this lgty?

jbergstroem commented:
I haven't tried it but the change LGTM.

Fishrock123 commented:
Rebased and updated, tested, can confirm it still works.

Unfortunately npm's tests still require a global node for now. ...unless that can be passed via env variable?

Fishrock123 commented on tools/test-npm.sh:
```diff
@@ -29,9 +29,10 @@ export npm_config_prefix="npm-prefix"
+
+../$NODE cli.js install --ignore-scripts
+../$NODE cli.js run-script test-legacy
+../$NODE cli.js run-script test
```
Actually on second thought this always used the local node to run cli.js anyways because these are relative links.

Still, I think this PR is better practice given our makefile setup.

Fishrock123 commented on tools/test-npm.sh:
```diff
@@ -3,11 +3,11 @@
 
 # always change the working directory to the project's root directory
-cd $(dirname $0)/..
+cd $(dirname ${BASH_SOURCE[0]})/..
```
see: http://stackoverflow.com/a/9107028/1279026 (related comment linked below in the "get absolute path" part).

silverwind replied:
Seems zsh doesn't sport this variable unfortunately, There is an [alternative](http://stackoverflow.com/a/23259585/808699), but I'd rather avoid using shell-specific things.

Fishrock123 replied:
Ok, Will revert to using `$0`

silverwind replied:
Hmm, what issue were you trying to solve here exactly? Maybe there's an alternative.

Fishrock123 replied:
Nothing really, I just saw it was suggested in the stackoverflow, going to revert to using `$0`

Fishrock123 commented:
Updated, PTAL. 

It now creates a local `node` link and exports it to the used `$PATH`, this should run the tests using only the local node binary. 

Will also test on my assigned linux machine.

Fishrock123 commented on tools/test-npm.sh:
```diff
@@ -17,6 +17,19 @@ rm -rf test-npm
+
+## a hopefully temporary hack to use the local $NODE binary when calling node(1)
+# make a link to $NODE as test-npm/node
+ln -f $NODE test-npm/node
```
This may be an issue cross-platform.

My OS X `ls` cannot do relative symlinks (`-r`), but linux does not appear to like hard links.

silverwind commented:
I wonder if we have other bash-specific things in those test-npm scripts. Could be an explanation for the weird issues I was having with test-npm.

Fishrock123 commented:
PTAL: Updated to remove this bit, which I'll PR separately:

```bash
## a hopefully temporary hack to use the local $NODE binary when calling node(1)
# make a link to $NODE as test-npm/node
ln -f $NODE test-npm/node
# get the absolute path we are in -- see: http://stackoverflow.com/a/4774063/1279026
pushd `dirname $0` > /dev/null
PWD=`pwd -P`
popd > /dev/null
# export that link as preferred in the $PATH
export PATH=$PWD/test-npm:$PATH
##
```

Fishrock123 commented on tools/test-npm.sh:
```diff
@@ -17,6 +17,19 @@ rm -rf test-npm
+popd > /dev/null
+# export that link as preferred in the $PATH
+export PATH=$PWD:$PATH
+##
```
Ok I can confirm this works on both OS X and Linux (Ubuntu 15), it just means we have an extra thing to clean up in the project root, as seen at the bottom of this file.

Fishrock123 replied:
Looks like this fails on a more recent npm with these errors:

```
test/tap/lifecycle-path.js ............................ 2/4
  make sure the path is correct
  not ok exit code
    +++ found                                                          
    --- wanted                                                         
    -0                                                                 
    +1                                                                 
    compare: ===
    at:
      file: test/tap/lifecycle-path.js
      line: 55
      column: 7
    stack: |
      test/tap/lifecycle-path.js:55:7
      f (node_modules/once/once.js:17:25)
      ChildProcess.<anonymous> (test/common-tap.js:56:5)
      maybeClose (internal/child_process.js:764:16)
      Socket.<anonymous> (internal/child_process.js:319:11)
      Pipe._onclose (net.js:467:12)
  
  make sure the path is correct
  not ok should be equivalent
    +++ found                                                          
    --- wanted                                                         
     [                                                                 
    -  "{{ROOT}}/bin/node-gyp-bin"                                     
    -  "{{ROOT}}/test/tap/lifecycle-path/node_modules/.bin"            
    -  "/bin"                                                          
    -  "/usr/bin"                                                      
    +  "> ./node-bin/node print-path.js"                               
     ]                                                                 
    at:
      file: test/tap/lifecycle-path.js
      line: 74
      column: 7
    stack: |
      test/tap/lifecycle-path.js:74:7
      f (node_modules/once/once.js:17:25)
      ChildProcess.<anonymous> (test/common-tap.js:56:5)
      maybeClose (internal/child_process.js:764:16)
      Socket.<anonymous> (internal/child_process.js:319:11)
      Pipe._onclose (net.js:467:12)
```

From this test: https://github.com/npm/npm/blob/master/test/tap/lifecycle-path.js

@othiym23 is this worth the effort to hack this in before npm's tests work without a global node? It seems like this test is relying on the location of the node binary, is that something that can / should be fixed?

Fishrock123 commented:
@silverwind / @jbergstroem ptal again. :)

Fishrock123 commented:
Anyone able to review? :D

evanlucas commented:
Hmm, getting some failures...not sure if it is related to the most recent changes you made or not though...

https://cloudup.com/cmC7mdzkL4x

Fishrock123 commented:
@evanlucas what OS? Those look like npm failures. If my script is broken on your system the tests will almost certainly fail to run. :P

evanlucas commented:
Yosemite. Haven't tried running them standalone yet though

Fishrock123 commented:
Uh oh, the same two tests are failing for me now when rebased on master, but only with this patch? How is that possible... I'll have to investigate.

silverwind commented on tools/test-npm.sh:
```diff
@@ -5,13 +5,26 @@ set -e
+PWD=`pwd -P`
+popd > /dev/null
+# export that link as preferred in the $PATH
+export PATH=$PWD:$PATH
```
I think the original PATH should be restored after the test run? (Maybe through saving it to a temp var)

Fishrock123 replied:
@silverwind you mean after the legacy tests? I could try that.

silverwind replied:
I mean having `PATH=$PWD:$PATH` will prepend the directory on each run of test-npm.sh, resulting in PATH getting longer and longer.

Fishrock123 replied:
Hmmm, nope. It appears the tests themselves depend on a global `node` (from the tap module, I think?)

Since running the test requires it, and causes it to fail, that's not a possibility.

silverwind replied:
Not sure you understand what I mean. My issue is with the global shell's PATH being in an modified state after this script finishes executing (global node now pointing to the built one), and I think we should restore PATH after this script is done.

silverwind replied:
Maybe just removing the `export` is sufficient so all child scripts get the correct binary and we avoid touching the global PATH.

silverwind commented on tools/test-npm.sh:
```diff
@@ -5,13 +5,26 @@ set -e
-# it should point to either {./}node or {./}node.exe, depending on the platform
-if [ -z $NODE_EXE ]; then
+# pass a $NODE environment variable from something like Makefile
+# it should point to either ./{node} or ./{node}.exe, depending on the platform
```
What's the significance of the braces in these comments, why not just write `./node`?

Fishrock123 replied:
Because in our case it it actually `./iojs`.

silverwind replied:
How about just `./iojs or ./iojs.exe` then?

Fishrock123 commented:
Updated to remove the $PATH hack. I'll try working on it separately. PTAL please, would like to land this haha.

silverwind commented:
Looks pretty straightforward now. LGTM

jbergstroem commented:
LGTM
